src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -5874,6 +5906,62 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                         "_lpython_str_startswith", loc);
             }
             return;
+        } else if (attr_name == "endswith") {
+            /* 
+                str.endswith(suffix)     ---->  
+                Return True if the string ends with the specified suffix, otherwise return False.
+            */
+
+            if (args.size() != 1) {
+                throw SemanticError("str.endswith() takes one arguments", loc);
+            }
+
+            ASR::expr_t *arg_suffix = args[0].m_value;
+            ASR::ttype_t *arg_suffix_type = ASRUtils::expr_type(arg_suffix);
+            if (!ASRUtils::is_character(*arg_suffix_type)) {
+                throw SemanticError("str.endswith() takes one arguments of type: str", arg_suffix->base.loc);
+            }
+            
+            if (ASRUtils::expr_value(arg_suffix) != nullptr) {
+                /*
+                    Invoked when Suffix argument is provided as a constant string
+                */
+                ASR::StringConstant_t* suffix_constant = ASR::down_cast<ASR::StringConstant_t>(arg_suffix);
+                std::string suffix = suffix_constant->m_s;
+                
+                bool res = true;
+                if (suffix.size() > s_var.size()) 
+                    res = false;
+                else 
+                    res = std::equal(suffix.rbegin(), suffix.rend(), s_var.rbegin());
+                
+                tmp = ASR::make_LogicalConstant_t(al, loc, res, 
+                    ASRUtils::TYPE(ASR::make_Logical_t(al, loc, 4, nullptr, 0)));
```

Comment:
> This should be refactored into IntrinsicFunction. Just add a new function. We will have thousands of such functions eventually.

How can I go on with refactoring of code to Intrinsic Functions? Can you give a head start? Because what I understood from @Thirumalai-Shaktivel is that we have to sync between LPython and LFortran for that. Can you give some pointers here?

Reply:
Yes, we have to sync first. Until then, you can implement this in LFortran itself, in libasr.

Reply:
Where in libasr? Can you give an example of one such addition in LFortran? Maybe a previous PR or something.

Reply:
in src/libasr.

Just look into merged PRs and search IntrinsicFunction. See also https://github.com/lcompilers/lpython/pull/1616 that is porting it to LPython.